girder_worker/app.py

```diff
@@ -109,35 +117,54 @@ def girder_before_task_publish(sender=None, body=None, exchange=None,
-                **{'title': headers.get('girder_job_title', Task._girder_job_title),
-                   'type': headers.get('girder_job_type', Task._girder_job_type),
-                   'handler': headers.get('girder_job_handler', Task._girder_job_handler),
-                   'public': headers.get('girder_job_public', Task._girder_job_public),
+                **{'title': headers.pop('girder_job_title', Task._girder_job_title),
+                   'type': headers.pop('girder_job_type', Task._girder_job_type),
+                   'handler': headers.pop('girder_job_handler', Task._girder_job_handler),
+                   'public': headers.pop('girder_job_public', Task._girder_job_public),
                    'user': user,
                    'args': task_args,
                    'kwargs': task_kwargs,
                    'otherFields': dict(celeryTaskId=headers['id'],
-                                       **headers.get('girder_job_other_fields',
+                                       **headers.pop('girder_job_other_fields',
                                                      Task._girder_job_other_fields))})
-            # If we don't have a token from girder_token kwarg,  use
-            # the job token instead. Otherwise no token
-            if token is None:
-                token = job.get('token', None)
 
-            headers['jobInfoSpec'] = utils.jobInfoSpec(job, token)
-            headers['apiUrl'] = utils.getWorkerApiUrl()
+            headers['jobInfoSpec'] = utils.jobInfoSpec(job)
 
         except ImportError:
             # TODO: Check for self.job_manager to see if we have
             #       tokens etc to contact girder and create a job model
             #       we may be in a chain or a chord or some-such
             pass
 
+    if 'girder_api_url' not in headers:
+        try:
+            from girder.plugins.worker import utils
+            headers['girder_api_url'] = utils.getWorkerApiUrl()
+        except ImportError:
+            # TODO: handle situation where girder_worker is producing the message
+            #       Note - this may not come up at all depending on how we pass
+            #       girder_api_url through to the next task (e.g. in the context
+            #       of chaining events)
+            pass
+
+    if 'girder_client_token' not in headers:
+        try:
+            from girder.utility.model_importer import ModelImporter
+            headers['girder_client_token'] = ModelImporter.model('token').createToken()
+        except ImportError:
+            # TODO: handle situation where girder_worker is producing the message
+            #       Note - this may not come up at all depending on how we pass
+            #       girder_token through to the next task (e.g. in the context
+            #       of chaining events)
+            pass
+
+    # Finally,  remove all special_options from headers
+    for key in Task.special_options:
+        headers.pop(key)
```

Comment:
I could go either way on it really - i've leaned towards no because i think its better to have the job model in girder be the oracle of this data (rather than reference the headers of the amqp message inside the task).  The nice thing is,  if there is a compelling reason to keep something in the headers it just needs to be moved from ```reserved_options```  to ```reserved_headers``` inside the ```girder_worker.app.Task``` class